graphqlapi/src/main/java/graphql/kickstart/servlet/CustomGraphQLServlet.java

```diff
@@ -0,0 +1,172 @@
+                }
+            }
+            handlerMap = map;
+        }
+    }
+
+    /**
+     * Populate a default keyspace to allow for omitting the keyspace from the path of requests.
+     *
+     * @param dataStore
+     */
+    private void populateDefaultKeyspace(DataStore dataStore) {
+        try {
+            CompletableFuture<ResultSet> query = dataStore.query(
+                    "select writetime(durable_writes) as wt, keyspace_name from system_schema.keyspaces",
+                    Optional.of(ConsistencyLevel.LOCAL_QUORUM));
+
+            ResultSet resultSet = query.get();
+
+            Column writetimeColumn = Column.create("wt", Column.Type.Counter);
+
+            // Grab the oldest, non-system keyspace to use as default.
+            Optional<Row> first = resultSet.rows().stream()
+                    .filter(r -> r.has("wt"))
+                    .filter(r -> r.getLong(writetimeColumn) > 0)
+                    .filter(r -> !r.getString("keyspace_name").equals("system") &&
+                            !r.getString("keyspace_name").equals("system_schema") &&
+                            !r.getString("keyspace_name").equals("data_endpoint_auth"))
+                    .min(Comparator.comparing(r -> r.getLong(writetimeColumn)));
+
+            DEFAULT_KEYSPACE = first.map(row -> row.getString("keyspace_name")).orElse(null);
+        } catch (Exception e) {
+            log.warn("Unable to get default keyspace", e);
+        }
+    }
+
+    protected void doGet(HttpServletRequest req, HttpServletResponse resp) {
+        this.init();
+
+        this.doRequestAsync(req, resp, this.handlerMap.get(getKeyspace(req)));
+    }
+
+    public String getKeyspace(HttpServletRequest request) {
+        String path = request.getPathInfo();
+        if (path == null) {
+            return DEFAULT_KEYSPACE;
+        }
+
+        return path.substring(1).toLowerCase();
+    }
+
+    protected void doPost(HttpServletRequest req, HttpServletResponse resp) {
+        this.init();
+
+        this.doRequestAsync(req, resp, this.handlerMap.get(getKeyspace(req)));
+    }
+
+    private void doRequestAsync(HttpServletRequest request, HttpServletResponse response, HttpRequestHandler handler) {
+        if (handler == null) {
+            response.setStatus(404);
+            throw new RuntimeException("Could not find keyspace");
+        }
+        if (this.isAsyncServletModeEnabled) {
+            AsyncContext asyncContext = request.startAsync(request, response);
+            HttpServletRequest asyncRequest = (HttpServletRequest)asyncContext.getRequest();
+            HttpServletResponse asyncResponse = (HttpServletResponse)asyncContext.getResponse();
+            this.asyncExecutor.execute(() -> {
+                this.doRequest(asyncRequest, asyncResponse, handler, asyncContext);
+            });
+        } else {
+            this.doRequest(request, response, handler, (AsyncContext)null);
+        }
```

Comment:
I'm not convinced by this async mode. It just schedules the exact same task on another executor. Granted, this frees up the HTTP thread to process another query, but that query will end up scheduling its own task on `asyncExecutor`, so we're not addressing the contention, only moving it one thread pool away.

Async works well when we deal with an async backend, for example if we used `DataStore` directly, we could complete the HTTP response in a callback on the `CompletionStage` returned by `query()`.

And in fact it looks like GraphQL supports that. From [the docs](https://www.graphql-java.com/documentation/v11/execution/):

> By default the “query” execution strategy is graphql.execution.AsyncExecutionStrategy which will dispatch each field as CompletableFuture objects and not care which ones complete first. This strategy allows for the most performant execution.
> 
> **The data fetchers invoked can themselves return CompletionStage values and this will create fully asynchronous behaviour.**

If I understand that correctly, if our data fetchers return futures, all we'd have to do in `doGet` and `doPost` is invoke `handler.handle(request, response)` and let it do its magic.

Reply:
Upon further investigation, GraphQL supports async but graphql-java-servlet doesn't 😢 . It blocks on the query result in [GraphQLInvoker](https://github.com/graphql-java-kickstart/graphql-java-servlet/blob/master/graphql-java-kickstart/src/main/java/graphql/kickstart/execution/GraphQLInvoker.java#L38).

There's even a pending feature request for that: graphql-java-kickstart/graphql-java-servlet#259.

I'll check if I can contribute a PR. In the meantime, I suggest we still remove `asyncExecutor`, because as I was explaining previously I don't think it really improves performance.